File: tools/artifact/schema.py
"""V3 directory records and structural validation, using only the standard library."""

from __future__ import annotations

from dataclasses import asdict, dataclass
import json
from math import prod
from typing import ClassVar, Mapping, Sequence, TypeAlias

from .layouts import align_up, encoded_size, get_layout

U64_MAX = (1 << 64) - 1
RAW_BYTES_V1 = "raw_bytes_v1"
ACTIVATION_POLICIES = frozenset(("A16Only", "AllowA8", "AllowA4"))


class ArtifactError(ValueError):
    """A directory or file does not satisfy the NInfer v3 contract."""


def integer(value: object, label: str, *, positive: bool = False) -> int:
    minimum = 1 if positive else 0
    if type(value) is not int or not minimum <= value <= U64_MAX:
        raise ArtifactError(f"{label}: expected integer in [{minimum},{U64_MAX}]")
    return value


def identifier(value: object, label: str) -> str:
    if not isinstance(value, str) or not value or "\0" in value:
        raise ArtifactError(f"{label}: expected nonempty string without NUL")
    try:
        value.encode("utf-8")
    except UnicodeEncodeError as error:
        raise ArtifactError(f"{label}: invalid UTF-8 string") from error
    return value


def shape(value: object, label: str) -> tuple[int, ...]:
    if not isinstance(value, (list, tuple)) or len(value) > 16:
        raise ArtifactError(f"{label}: expected shape with rank 0 through 16")
    result = tuple(integer(dim, label, positive=True) for dim in value)
    integer(prod(result), f"{label} element count", positive=True)
    return result


def members(
    value: object,
    required: set[str],
    optional: set[str],
    label: str,
) -> dict:
    if not isinstance(value, dict):
        raise ArtifactError(f"{label}: expected object")
    missing = required - value.keys()
    extra = value.keys() - required - optional
    if missing or extra:
        raise ArtifactError(
            f"{label}: missing {sorted(missing)}, unknown {sorted(extra)}"
        )
    return value


@dataclass(frozen=True, slots=True)
class TensorSpec:
    id: str
    shape: tuple[int, ...]
    format: str
    layout: str


@dataclass(frozen=True, slots=True)
class ResourceSpec:
    id: str
    bytes: int
    encoding: str = RAW_BYTES_V1


@dataclass(frozen=True, slots=True)
class TensorObject:
    id: str
    shape: tuple[int, ...]
    format: str
    layout: str
    offset: int
    bytes: int
    kind: ClassVar[str] = "tensor"

    def to_json(self) -> dict:
        return {**asdict(self), "shape": list(self.shape), "kind": self.kind}


@dataclass(frozen=True, slots=True)
class ResourceObject:
    id: str
    encoding: str
    offset: int
    bytes: int
    kind: ClassVar[str] = "resource"

    def to_json(self) -> dict:
        return {**asdict(self), "kind": self.kind}


ObjectSpec: TypeAlias = TensorSpec | ResourceSpec
ArtifactObject: TypeAlias = TensorObject | ResourceObject


@dataclass(frozen=True, slots=True)
class FileRecord:
    path: str | None
    payload_bytes: int


@dataclass(frozen=True, slots=True)
class Directory:
    components: dict
    objects: tuple[ArtifactObject, ...]
    bindings: dict
    uses: tuple[dict, ...]
    files: tuple[FileRecord, ...]
    metadata: dict
    provenance: dict

    @property
    def payload_bytes(self) -> int:
        return sum(file.payload_bytes for file in self.files)

    def to_json(self) -> dict:
        return {
            "components": self.components,
            "objects": [obj.to_json() for obj in self.objects],
            "bindings": self.bindings,
            "uses": list(self.uses),
            "files": [asdict(file) for file in self.files],
            "metadata": self.metadata,
            "provenance": self.provenance,
        }


def validate_encoding(obj: ArtifactObject) -> None:
    """Interpret an object only when a consumer requests its encoding."""
    if isinstance(obj, ResourceObject):
        if obj.encoding != RAW_BYTES_V1:
            raise ArtifactError(
                f"{obj.id}: unsupported resource encoding {obj.encoding!r}"
            )
        return
    try:
        size = encoded_size(obj.layout, obj.format, obj.shape)
        alignment = get_layout(obj.layout).alignment
    except (TypeError, ValueError) as error:
        raise ArtifactError(f"{obj.id}: {error}") from error
    integer(size, f"{obj.id} encoded size", positive=True)
    if obj.bytes != size or obj.offset % alignment:
        raise ArtifactError(
            f"{obj.id}: expected {size} encoded bytes and {alignment}-byte alignment, "
            f"got bytes={obj.bytes}, offset={obj.offset}"
        )


def plan_objects(specs: Sequence[ObjectSpec]) -> tuple[ArtifactObject, ...]:
    result = []
    seen = set()
    offset = 0
    for spec in specs:
        name = identifier(spec.id, "object id")
        if name in seen:
            raise ArtifactError(f"duplicate object id {name!r}")
        seen.add(name)
        if isinstance(spec, TensorSpec):
            dims = shape(spec.shape, name)
            try:
                start = align_up(offset, get_layout(spec.layout).alignment)
                size = encoded_size(spec.layout, spec.format, dims)
            except (TypeError, ValueError) as error:
                raise ArtifactError(f"{name}: {error}") from error
            obj = TensorObject(name, dims, spec.format, spec.layout, start, size)
        elif isinstance(spec, ResourceSpec):
            obj = ResourceObject(name, spec.encoding, offset, spec.bytes)
        else:
            raise TypeError(
                f"expected TensorSpec or ResourceSpec, got {type(spec).__name__}"
            )
        integer(obj.bytes, f"{name} bytes", positive=True)
        offset = integer(obj.offset + obj.bytes, f"{name} end")
        validate_encoding(obj)
        result.append(obj)
    if not result:
        raise ArtifactError("objects must not be empty")
    return tuple(result)


def _parse_object(value: object) -> ArtifactObject:
    if not isinstance(value, dict):
        raise ArtifactError("object descriptor must be an object")
    common = {"id", "kind", "offset", "bytes"}
    kind = value.get("kind")
    if kind == "tensor":
        members(value, common | {"shape", "format", "layout"}, set(), "tensor")
    elif kind == "resource":
        members(value, common | {"encoding"}, set(), "resource")
    else:
        raise ArtifactError(f"unsupported object kind {kind!r}")
    name = identifier(value["id"], "object id")
    offset = integer(value["offset"], f"{name} offset")
    size = integer(value["bytes"], f"{name} bytes", positive=True)
    integer(offset + size, f"{name} end")
    if kind == "resource":
        return ResourceObject(name, identifier(value["encoding"], name), offset, size)
    return TensorObject(
        name,
        shape(value["shape"], name),
        identifier(value["format"], name),
        identifier(value["layout"], name),
        offset,
        size,
    )


def binding_parts(
    binding: object,
    objects: Mapping[str, ArtifactObject],
    label: str = "binding",
) -> tuple[tuple[str, int, int], ...]:
    """Resolve flat logical element ranges without interpreting tensor codecs."""
    if not isinstance(binding, dict):
        raise ArtifactError(f"{label}: expected Binding object")

    def tensor(name: object) -> TensorObject:
        name = identifier(name, label)
        obj = objects.get(name)
        if not isinstance(obj, TensorObject):
            raise ArtifactError(f"{label}: {name!r} does not reference a tensor")
        return obj

    if set(binding) == {"object"}:
        obj = tensor(binding["object"])
        return ((obj.id, 0, prod(obj.shape)),)
    members(binding, {"parts"}, set(), label)
    values = binding["parts"]
    if not isinstance(values, list) or not values:
        raise ArtifactError(f"{label}: parts must be a nonempty array")
    result = []
    total = 0
    for part in values:
        members(part, {"object", "range"}, set(), label)
        obj = tensor(part["object"])
        bounds = part["range"]
        if not isinstance(bounds, list) or len(bounds) != 2:
            raise ArtifactError(f"{label}: range must contain two integers")
        begin, end = (integer(v, label) for v in bounds)
        if not begin < end <= prod(obj.shape):
            raise ArtifactError(
                f"{label}: range {bounds} is outside {obj.id} {obj.shape}"
            )
        total = integer(total + end - begin, f"{label} element count", positive=True)
        result.append((obj.id, begin, end))
    return tuple(result)


def _files(values: object, entry_name: str | None) -> tuple[FileRecord, ...]:
    if not isinstance(values, list) or not values:
        raise ArtifactError("files must be a nonempty array")
    result = []
    seen = {entry_name} if entry_name is not None else set()
    total = 0
    for index, value in enumerate(values):
        members(value, {"path", "payload_bytes"}, set(), f"files[{index}]")
        name = value["path"]
        if index == 0:
            if name is not None:
                raise ArtifactError("files[0].path must be null")
        else:
            name = identifier(name, f"files[{index}].path")
            if name in (".", "..") or "/" in name or "\\" in name or name in seen:
                raise ArtifactError(
                    f"files[{index}]: invalid or duplicate sibling filename {name!r}"
                )
            seen.add(name)
        size = integer(
            value["payload_bytes"], f"files[{index}] payload bytes", positive=True
        )
        total = integer(total + size, "total payload bytes", positive=True)
        result.append(FileRecord(name, size))
    return tuple(result)


def parse_directory(value: object, *, entry_name: str | None = None) -> Directory:
    root = members(
        value,
        {"components", "objects", "bindings", "uses", "files"},
        {"metadata", "provenance"},
        "directory",
    )
    files = _files(root["files"], entry_name)
    raw_objects = root["objects"]
    if not isinstance(raw_objects, list) or not raw_objects:
        raise ArtifactError("objects must be a nonempty array")
    objects = tuple(_parse_object(item) for item in raw_objects)
    index = {}
    previous_end = 0
    payload_bytes = sum(file.payload_bytes for file in files)
    for obj in objects:
        if obj.id in index:
            raise ArtifactError(f"duplicate object id {obj.id!r}")
        if obj.offset < previous_end or obj.offset + obj.bytes > payload_bytes:
            raise ArtifactError(
                f"{obj.id}: object ranges overlap, are out of order, or exceed payload"
            )
        previous_end = obj.offset + obj.bytes
        index[obj.id] = obj

    components = root["components"]
    if not isinstance(components, dict) or "text" not in components:
        raise ArtifactError("components must contain text")
    for name, component in components.items():
        identifier(name, "component id")
        members(component, {"config"}, {"target", "resources", "proposal"}, name)
        if not isinstance(component["config"], dict):
            raise ArtifactError(f"{name}: config must be an object")
        if "target" in component:
            target = identifier(component["target"], f"{name} target")
            if target not in components:
                raise ArtifactError(f"{name}: missing target {target!r}")
        resources = component.get("resources", {})
        if not isinstance(resources, dict):
            raise ArtifactError(f"{name}: resources must be an object")
        for role, object_id in resources.items():
            identifier(role, f"{name} resource role")
            object_id = identifier(object_id, f"{name}/{role}")
            if not isinstance(index.get(object_id), ResourceObject):
                raise ArtifactError(
                    f"{name}/{role}: missing resource object {object_id!r}"
                )
        if "proposal" in component:
            if name != "text":
                raise ArtifactError("only text may declare proposal")
            proposal = members(component["proposal"], {"domain"}, {"rows"}, "proposal")
            if proposal["domain"] == "full":
                if "rows" in proposal:
                    raise ArtifactError("full proposal must omit rows")
            elif proposal["domain"] == "indexed":
                integer(proposal.get("rows"), "proposal rows", positive=True)
            else:
                raise ArtifactError("proposal domain must be full or indexed")

    bindings = root["bindings"]
    if not isinstance(bindings, dict):
        raise ArtifactError("bindings must be an object")
    for name, binding in bindings.items():
        identifier(name, "parameter")
        binding_parts(binding, index, name)
    uses = root["uses"]
    if not isinstance(uses, list):
        raise ArtifactError("uses must be an array")
    seen_uses = set()
    for use in uses:
        members(
            use, {"parameter", "input"}, {"activation_policy", "auxiliaries"}, "Use"
        )
        name = identifier(use["parameter"], "Use parameter")
        source = identifier(use["input"], f"{name} Use input")
        if name not in bindings or (name, source) in seen_uses:
            raise ArtifactError(f"{name}@{source}: missing parameter or duplicate Use")
        seen_uses.add((name, source))
        if "activation_policy" in use:
            policy = identifier(use["activation_policy"], f"{name} activation policy")
            if policy not in ACTIVATION_POLICIES:
                raise ArtifactError(f"{name}: unsupported activation policy {policy!r}")
        auxiliaries = use.get("auxiliaries", {})
        if not isinstance(auxiliaries, dict):
            raise ArtifactError(f"{name}: auxiliaries must be an object")
        for role, binding in auxiliaries.items():
            identifier(role, f"{name} auxiliary role")
            binding_parts(binding, index, f"{name}@{source}/{role}")
    metadata = root.get("metadata", {})
    provenance = root.get("provenance", {})
    if not isinstance(metadata, dict) or not isinstance(provenance, dict):
        raise ArtifactError("metadata and provenance must be objects")
    if "name" in metadata:
        identifier(metadata["name"], "metadata.name")
    return Directory(
        components, objects, bindings, tuple(uses), files, metadata, provenance
    )


def _unique_members(pairs: list[tuple[str, object]]) -> dict:
    result = {}
    for key, value in pairs:
        if key in result:
            raise ArtifactError(f"duplicate JSON member {key!r}")
        result[key] = value
    return result


def _nonfinite(value: str) -> None:
    raise ArtifactError(f"nonfinite JSON number {value}")


def decode_directory(data: bytes, *, entry_name: str | None = None) -> Directory:
    try:
        value = json.loads(
            data.decode("utf-8"),
            object_pairs_hook=_unique_members,
            parse_constant=_nonfinite,
        )
    except (UnicodeError, json.JSONDecodeError) as error:
        raise ArtifactError(f"invalid directory JSON: {error}") from error
    return parse_directory(value, entry_name=entry_name)


def encode_directory(directory: Directory | dict) -> bytes:
    value = directory.to_json() if isinstance(directory, Directory) else directory
    try:
        return json.dumps(
            value, ensure_ascii=False, allow_nan=False, separators=(",", ":")
        ).encode("utf-8")
    except (UnicodeError, ValueError, TypeError) as error:
        raise ArtifactError(f"cannot encode directory JSON: {error}") from error

File: src/artifact/schema.cpp
#include "artifact/schema.h"

#include "core/weight_view.h"

#include <algorithm>
#include <set>
#include <unordered_set>

namespace ninfer::artifact {

std::uint64_t checked_add(std::uint64_t a, std::uint64_t b, std::string_view label) {
    if (b > std::numeric_limits<std::uint64_t>::max() - a) {
        throw ArtifactError(std::string(label) + " overflows u64");
    }
    return a + b;
}

std::uint64_t checked_mul(std::uint64_t a, std::uint64_t b, std::string_view label) {
    if (a && b > std::numeric_limits<std::uint64_t>::max() / a) {
        throw ArtifactError(std::string(label) + " overflows u64");
    }
    return a * b;
}

std::uint64_t align_up(std::uint64_t value, std::uint64_t alignment, std::string_view label) {
    if (!alignment) { throw ArtifactError(std::string(label) + ": zero alignment"); }
    return checked_add(value, alignment - 1, label) / alignment * alignment;
}

void require_members(const Json& value, std::initializer_list<std::string_view> required,
                     std::initializer_list<std::string_view> optional, std::string_view label) {
    if (!value.is_object()) { throw ArtifactError(std::string(label) + " must be an object"); }
    for (const auto key : required) {
        if (!value.contains(key)) {
            throw ArtifactError(std::string(label) + ": missing " + std::string(key));
        }
    }
    for (const auto& [key, unused] : value.items()) {
        if (std::find(required.begin(), required.end(), key) == required.end() &&
            std::find(optional.begin(), optional.end(), key) == optional.end()) {
            throw ArtifactError(std::string(label) + ": unknown member " + key);
        }
    }
}

std::string require_id(const Json& value, std::string_view label) {
    if (!value.is_string()) { throw ArtifactError(std::string(label) + " must be a string"); }
    auto result = value.get<std::string>();
    if (result.empty() || result.find('\0') != std::string::npos) {
        throw ArtifactError(std::string(label) + " must be nonempty without NUL");
    }
    return result;
}

std::uint64_t require_u64(const Json& value, std::string_view label, bool positive) {
    if (!value.is_number_integer() ||
        (!value.is_number_unsigned() && value.get<std::int64_t>() < 0)) {
        throw ArtifactError(std::string(label) + " must be a nonnegative integer");
    }
    const auto result = value.get<std::uint64_t>();
    if (positive && !result) { throw ArtifactError(std::string(label) + " must be positive"); }
    return result;
}

const std::string& object_id(const Object& object) {
    return std::visit([](const auto& item) -> const std::string& { return item.id; }, object);
}

std::uint64_t object_offset(const Object& object) {
    return std::visit([](const auto& item) { return item.offset; }, object);
}

std::uint64_t object_bytes(const Object& object) {
    return std::visit([](const auto& item) { return item.bytes; }, object);
}

const Object& Directory::object(ObjectHandle handle) const {
    if (handle.index >= objects.size()) { throw ArtifactError("invalid object handle"); }
    return objects[handle.index];
}

const TensorObject& Directory::tensor(ObjectHandle handle) const {
    const auto* result = std::get_if<TensorObject>(&object(handle));
    if (!result) { throw ArtifactError("object does not reference a tensor"); }
    return *result;
}

const Component& Directory::component(std::string_view name) const {
    const auto found = components.find(name);
    if (found == components.end()) {
        throw ArtifactError("missing component " + std::string(name));
    }
    return found->second;
}

Json parse_json(std::string_view text, std::string_view label) {
    std::vector<std::unordered_set<std::string>> member_stack;
    const auto callback = [&](int, Json::parse_event_t event, Json& parsed) {
        if (event == Json::parse_event_t::object_start) {
            member_stack.emplace_back();
        } else if (event == Json::parse_event_t::key) {
            const auto& key = parsed.get_ref<const std::string&>();
            if (!member_stack.back().insert(key).second) {
                throw ArtifactError(std::string(label) + ": duplicate JSON member " + key);
            }
        } else if (event == Json::parse_event_t::object_end) {
            member_stack.pop_back();
        }
        return true;
    };
    try {
        return Json::parse(text.begin(), text.end(), callback);
    } catch (const Json::exception& error) {
        throw ArtifactError(std::string(label) + ": " + error.what());
    }
}

namespace {

Shape parse_shape(const Json& value, std::string_view label) {
    if (!value.is_array() || value.size() > 16) {
        throw ArtifactError(std::string(label) + ": shape must have rank 0 through 16");
    }
    Shape out;
    std::uint64_t elements = 1;
    for (const auto& item : value) {
        const auto dim = require_u64(item, label, true);
        elements       = checked_mul(elements, dim, label);
        out.push_back(dim);
    }
    return out;
}

ObjectHandle reference(const Directory& directory, const Json& value, std::string_view label,
                       bool tensor) {
    const auto name  = require_id(value, label);
    const auto found = directory.object_index.find(name);
    if (found == directory.object_index.end()) {
        throw ArtifactError(std::string(label) + ": missing object " + name);
    }
    if (std::holds_alternative<TensorObject>(directory.object(found->second)) != tensor) {
        throw ArtifactError(std::string(label) + ": wrong object kind for " + name);
    }
    return found->second;
}

Binding parse_binding(const Directory& directory, const Json& value, std::string_view label) {
    if (!value.is_object()) {
        throw ArtifactError(std::string(label) + ": expected Binding object");
    }
    Binding out;
    if (value.size() == 1 && value.contains("object")) {
        const auto handle  = reference(directory, value.at("object"), label, true);
        const auto& object = directory.tensor(handle);
        out.whole_object   = true;
        out.elements       = weight_element_count(object.shape);
        out.parts.push_back({handle, 0, out.elements});
        return out;
    }
    require_members(value, {"parts"}, {}, label);
    const auto& parts = value.at("parts");
    if (!parts.is_array() || parts.empty()) {
        throw ArtifactError(std::string(label) + ": parts must be a nonempty array");
    }
    for (const auto& part : parts) {
        require_members(part, {"object", "range"}, {}, label);
        const auto handle  = reference(directory, part.at("object"), label, true);
        const auto& bounds = part.at("range");
        if (!bounds.is_array() || bounds.size() != 2) {
            throw ArtifactError(std::string(label) + ": range must contain two integers");
        }
        const auto begin = require_u64(bounds[0], label);
        const auto end   = require_u64(bounds[1], label);
        if (begin >= end || end > weight_element_count(directory.tensor(handle).shape)) {
            throw ArtifactError(std::string(label) + ": Part exceeds source logical elements");
        }
        out.elements = checked_add(out.elements, end - begin, label);
        out.parts.push_back({handle, begin, end});
    }
    return out;
}

void parse_files(Directory& out, const Json& files, std::string_view entry_name) {
    if (!files.is_array() || files.empty()) {
        throw ArtifactError("files must be a nonempty array");
    }
    std::set<std::string, std::less<>> names{std::string(entry_name)};
    for (const auto& value : files) {
        require_members(value, {"path", "payload_bytes"}, {}, "file");
        FileRecord file;
        if (out.files.empty()) {
            if (!value.at("path").is_null()) { throw ArtifactError("files[0].path must be null"); }
        } else {
            file.path        = require_id(value.at("path"), "file path");
            const auto& path = *file.path;
            if (path == "." || path == ".." || path.find('/') != std::string::npos ||
                path.find('\\') != std::string::npos || !names.insert(path).second) {
                throw ArtifactError("invalid or duplicate sibling filename: " + path);
            }
        }
        file.payload_bytes = require_u64(value.at("payload_bytes"), "file payload bytes", true);
        file.logical_begin = out.payload_bytes;
        out.payload_bytes =
            checked_add(out.payload_bytes, file.payload_bytes, "total payload bytes");
        out.files.push_back(std::move(file));
    }
}

void parse_objects(Directory& out, const Json& objects) {
    if (!objects.is_array() || objects.empty()) {
        throw ArtifactError("objects must be a nonempty array");
    }
    std::uint64_t previous_end = 0;
    for (const auto& value : objects) {
        if (!value.is_object() || !value.contains("kind")) {
            throw ArtifactError("invalid object descriptor");
        }
        const auto kind = require_id(value.at("kind"), "object kind");
        if (kind == "tensor") {
            require_members(value, {"id", "kind", "shape", "format", "layout", "offset", "bytes"},
                            {}, "tensor");
        } else if (kind == "resource") {
            require_members(value, {"id", "kind", "encoding", "offset", "bytes"}, {}, "resource");
        } else {
            throw ArtifactError("unknown object kind " + kind);
        }
        auto id           = require_id(value.at("id"), "object id");
        const auto offset = require_u64(value.at("offset"), id);
        const auto bytes  = require_u64(value.at("bytes"), id, true);
        const auto end    = checked_add(offset, bytes, id);
        if (offset < previous_end || end > out.payload_bytes) {
            throw ArtifactError(id + ": object overlaps, is out of order or exceeds payload");
        }
        if (!out.object_index.emplace(id, ObjectHandle{out.objects.size()}).second) {
            throw ArtifactError("duplicate object id " + id);
        }
        previous_end = end;
        if (kind == "tensor") {
            out.objects.emplace_back(TensorObject{
                id, parse_shape(value.at("shape"), id), require_id(value.at("format"), id),
                require_id(value.at("layout"), id), offset, bytes});
        } else {
            out.objects.emplace_back(
                ResourceObject{id, require_id(value.at("encoding"), id), offset, bytes});
        }
    }
}

void parse_components(Directory& out, const Json& values) {
    if (!values.is_object() || !values.contains("text")) {
        throw ArtifactError("components must contain text");
    }
    for (const auto& [name, value] : values.items()) {
        (void)require_id(Json(name), "component id");
        require_members(value, {"config"}, {"target", "resources", "proposal"}, name);
        Component component;
        component.config = value.at("config");
        if (!component.config.is_object()) {
            throw ArtifactError(name + ": config must be an object");
        }
        if (value.contains("target")) {
            component.target = require_id(value.at("target"), name + " target");
            if (!values.contains(*component.target)) {
                throw ArtifactError(name + ": missing target " + *component.target);
            }
        }
        if (value.contains("resources")) {
            const auto& resources = value.at("resources");
            if (!resources.is_object()) {
                throw ArtifactError(name + ": resources must be an object");
            }
            for (const auto& [role, ref] : resources.items()) {
                (void)require_id(Json(role), "resource role");
                component.resources.emplace(role, reference(out, ref, name + "/" + role, false));
            }
        }
        if (value.contains("proposal")) {
            if (name != "text") { throw ArtifactError("only text may declare proposal"); }
            const auto& p = value.at("proposal");
            require_members(p, {"domain"}, {"rows"}, "proposal");
            const auto domain = require_id(p.at("domain"), "proposal domain");
            if (domain == "full" && !p.contains("rows")) {
                component.proposal = Proposal{};
            } else if (domain == "indexed" && p.contains("rows")) {
                component.proposal =
                    Proposal{true, require_u64(p.at("rows"), "proposal rows", true)};
            } else {
                throw ArtifactError(
                    "proposal must be full without rows or indexed with positive rows");
            }
        }
        out.components.emplace(name, std::move(component));
    }
}

void parse_uses(Directory& out, const Json& values) {
    if (!values.is_array()) { throw ArtifactError("uses must be an array"); }
    for (const auto& value : values) {
        require_members(value, {"parameter", "input"}, {"activation_policy", "auxiliaries"}, "Use");
        Use use;
        use.parameter    = require_id(value.at("parameter"), "Use parameter");
        use.input        = require_id(value.at("input"), "Use input");
        const auto label = use.parameter + "@" + use.input;
        if (!out.bindings.contains(use.parameter)) {
            throw ArtifactError(label + ": missing parameter");
        }
        if (value.contains("activation_policy")) {
            const auto policy = require_id(value.at("activation_policy"), label);
            if (policy == "A16Only") {
                use.activation_policy = ActivationPolicy::A16Only;
            } else if (policy == "AllowA8") {
                use.activation_policy = ActivationPolicy::AllowA8;
            } else if (policy == "AllowA4") {
                use.activation_policy = ActivationPolicy::AllowA4;
            } else {
                throw ArtifactError(label + ": unknown activation policy " + policy);
            }
        }
        if (value.contains("auxiliaries")) {
            const auto& auxiliaries = value.at("auxiliaries");
            if (!auxiliaries.is_object()) {
                throw ArtifactError(label + ": auxiliaries must be an object");
            }
            for (const auto& [role, binding] : auxiliaries.items()) {
                (void)require_id(Json(role), "auxiliary role");
                use.auxiliaries.emplace(role, parse_binding(out, binding, label + "/" + role));
            }
        }
        const auto key = std::pair{use.parameter, use.input};
        if (!out.uses.emplace(key, std::move(use)).second) {
            throw ArtifactError(label + ": duplicate Use");
        }
    }
}

} // namespace

Directory parse_directory(const Json& root, std::string_view entry_name) {
    require_members(root, {"components", "objects", "bindings", "uses", "files"},
                    {"metadata", "provenance"}, "directory");
    Directory out;
    parse_files(out, root.at("files"), entry_name);
    parse_objects(out, root.at("objects"));
    parse_components(out, root.at("components"));
    const auto& bindings = root.at("bindings");
    if (!bindings.is_object()) { throw ArtifactError("bindings must be an object"); }
    for (const auto& [name, value] : bindings.items()) {
        (void)require_id(Json(name), "parameter name");
        out.bindings.emplace(name, parse_binding(out, value, name));
    }
    parse_uses(out, root.at("uses"));
    if (root.contains("metadata")) { out.metadata = root.at("metadata"); }
    if (root.contains("provenance")) { out.provenance = root.at("provenance"); }
    if (!out.metadata.is_object() || !out.provenance.is_object()) {
        throw ArtifactError("metadata and provenance must be objects");
    }
    if (out.metadata.contains("name")) {
        (void)require_id(out.metadata.at("name"), "metadata.name");
    }
    return out;
}

} // namespace ninfer::artifact
